pkg/provisioner/dependency.go

```diff
@@ -191,3 +86,78 @@ func kind(tpl *bytes.Buffer, env v1alpha1.Environment) error {
 	}
 	return kind.Execute(tpl, env)
 }
+
+// DependencySolver is a struct that holds the dependency list
+type DependencyResolver struct {
+	Dependencies []ProvisionFunc
+	env          v1alpha1.Environment
+}
+
+func NewDependencies(env v1alpha1.Environment) *DependencyResolver {
+	return &DependencyResolver{
+		env: env,
+	}
+}
+
+func (d *DependencyResolver) withKubernetes() {
+	switch d.env.Spec.Kubernetes.KubernetesInstaller {
+	case "kubeadm":
+		d.Dependencies = append(d.Dependencies, functions["kubeadm"])
+	case "kind":
+		d.Dependencies = append(d.Dependencies, functions["kind"])
+	case "microk8s":
+		// reset the list to only include microk8s
+		d.Dependencies = nil
+		d.Dependencies = append(d.Dependencies, functions["microk8s"])
+	default:
+		// default to kubeadm if KubernetesInstaller is empty
+		d.Dependencies = append(d.Dependencies, functions["kubeadm"])
+	}
+}
+
+func (d *DependencyResolver) withContainerRuntime() {
+	switch d.env.Spec.ContainerRuntime.Name {
+	case "containerd":
+		d.Dependencies = append(d.Dependencies, functions["containerd"])
+	case "crio":
+		d.Dependencies = append(d.Dependencies, functions["crio"])
+	case "docker":
+		d.Dependencies = append(d.Dependencies, functions["docker"])
+	default:
+		// default to containerd if ContainerRuntime.Name is empty
+		d.Dependencies = append(d.Dependencies, functions["containerd"])
+	}
+}
+
+func (d *DependencyResolver) withContainerToolkit() {
+	d.Dependencies = append(d.Dependencies, functions["containerToolkit"])
+}
+
+func (d *DependencyResolver) withNVDriver() {
+	d.Dependencies = append(d.Dependencies, functions["nvdriver"])
+}
+
+// Resolve returns the dependency list in the correct order
+func (d *DependencyResolver) Resolve() []ProvisionFunc {
+	// Add NVDriver to the list
+	if d.env.Spec.NVIDIADriver.Install {
```

Comment:
I wonder if we can have a function map instead and we could loop through the keys of the map and invoke each function.

The keys of the function map here would be `["nvdriver", "container-runtime", "container-toolkit", "kubernetes"]`

Reply:
we need it to be an ordered list, to install dependencies in order.

Reply:
Sure. The keys can be in a slice which guarantees order and then you can invoke the function map as you iterate through the slice